File: archive/c/c-sharp/Zeckendorf.cs
using System;
using System.Collections.Generic;
using System.Globalization;

if (args is not [var raw] ||
    !long.TryParse(raw, out long n) ||
    n < 0)
{
    return ExitWithUsage();
}

if (n == 0)
    return 0;

ReadOnlySpan<long> fibs = GenerateFibs();
Span<long> buffer = stackalloc long[fibs.Length];

int count = Decompose(n, fibs, buffer);
Console.WriteLine(string.Join(", ", buffer[..count].ToArray()));

return 0;

static int Decompose(long n, ReadOnlySpan<long> fibs, Span<long> result)
{
    int count = 0;

    for (int i = fibs.Length - 1; i >= 0 && n > 0; i--)
    {
        long f = fibs[i];

        if (f <= n)
        {
            result[count++] = f;
            n -= f;
        }
    }

    return count;
}

static int ExitWithUsage()
{
    Console.WriteLine("Usage: please input a non-negative integer");
    return 1;
}

static long[] GenerateFibs()
{
    var list = new List<long> { 1, 2 };

    while (true)
    {
        long next = list[^1] + list[^2];
        if (next < 0 || next > long.MaxValue - list[^1])
            break;

        list.Add(next);
    }

    return list.ToArray();
}

File: archive/v/visual-basic/zeckendorf.vb
Public Module Zeckendorf

    Private ReadOnly Fibs As Long() = GenerateFibs()


    Private Function GenerateFibs() As Long()

        Dim fs As New List(Of Long) From {1L, 2L}

        While Long.MaxValue - fs(fs.Count - 1) >= fs(fs.Count - 2)
            fs.Add(fs(fs.Count - 1) + fs(fs.Count - 2))
        End While

        Return fs.ToArray()

    End Function


    Public Sub Main(args As String())

        Dim n As Long

        If args.Length = 0 OrElse Not Long.TryParse(args(0), n) OrElse n < 0 Then

            Console.WriteLine("Usage: please input a non-negative integer")
            Return

        End If

        If n = 0 Then Return

        Dim terms As New List(Of Long)

        Dim count As Integer = Decompose(n, terms)

        Console.WriteLine(String.Join(", ", terms.GetRange(0, count)))

    End Sub


    Private Function Decompose(n As Long, terms As List(Of Long)) As Integer

        Dim count As Integer = 0

        Dim i As Integer = Array.BinarySearch(Fibs, n)

        If i < 0 Then
            i = Not i - 1
        End If

        For idx As Integer = i To 0 Step -1

            If n <= 0 Then Exit For

            If Fibs(idx) <= n Then
                terms.Add(Fibs(idx))
                n -= Fibs(idx)
                count += 1
            End If

        Next

        Return count

    End Function

End Module